README
Freeside

Copyright (C) 2000,2001,2002,2003 Ivan Kohler
Copyright (C) 1999 Silicon Interactive Software Design
All rights reserved

    This program is free software; you can redistribute it and/or modify
    it under the terms of:

        a) the GNU General Public License as published by the Free
        Software Foundation; either version 2, or (at your option) any
        later version

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License more details.
  
    You should have received a copy of the GNU General Public
    License along with this program, in the file `GPL'; if not,
    write to the Free Software Foundation, Inc., 59 Temple Place - Suite
    330, Boston, MA 02111-1307, USA.

Freeside is a billing and administration package for Internet Service 
Providers.

The Freeside home page is at `http://www.sisd.com/freeside'.

The documentation is in `httemplate/docs'.

A mailing list for users is available.  Send a blank message to
<ivan-freeside-subscribe@sisd.com> to subscribe.

A mailing list for developers is available.  It is intended to be lower volume
and higher SNR than the users list.  Send a blank message to
<ivan-freeside-devel-subscribe@sisd.com> to subscribe.

Commercial support is available from Ivan Kohler <ivan@sisd.com>.  Requests for
free support sent to me directly will be ignored.  Please subscribe to the the
user mailing list to request free support!

Ivan Kohler <ivan-freeside_readme@420.am>



ChangeLog
2011-10-24 17:23  ivan

	* Makefile, httemplate/docs/about.html: 2.1.3

2011-10-24 17:19  ivan

	* FS/FS/: svc_acct.pm, svc_broadband.pm: fix error on advanced
	  broadband report when selecting a router, RT#14626

2011-10-24 17:04  ivan

	* FS/FS/cust_main/Billing.pm: make discount-show-always get along
	  with setup discounts, RT#11512

2011-10-24 15:55  ivan

	* FS/FS/part_pkg/discount_Mixin.pm: fix discounting setup when
	  recur < setup, RT#11512

2011-10-24 15:45  ivan

	* httemplate/browse/discount.html: fix browsing of unnamed
	  discounds

2011-10-24 14:20  ivan

	* FS/FS/part_pkg/voip_inbound.pm: fix display of duration on
	  inbound packages with use_duraiton, RT#14370

2011-10-23 18:15  ivan

	* FS/FS/cust_pkg_discount.pm, FS/FS/part_pkg/flat.pm,
	  httemplate/elements/order_pkg.js,
	  httemplate/elements/tr-select-discount.html,
	  httemplate/view/cust_main/order_pkg_link.html: fix discounts not
	  appearing for one-time charge packages, RT#13654

2011-10-23 16:09  ivan

	* httemplate/elements/menu.html: reports->packages->package
	  definitions now "by # active", RT#5501

2011-09-28 22:53  ivan

	* FS/FS/Record.pm: fix DBI binding for negative integers, shoudl
	  fix CCH update, RT#14243

2011-09-28 22:22  ivan

	* FS/FS/part_pkg_taxrate.pm: add additional debugging for failing
	  CCH update, RT#14243

2011-09-22 15:07  ivan

	* FS/FS/Conf_compat17.pm: whitespace

2011-09-14 13:48  ivan

	* FS/FS/ClientAPI/MyAccount.pm: add svc_label to ClientAPI
	  customer_info, RT#13656

2011-09-12 13:06  ivan

	* FS/FS/UID.pm: fix no REMOTE_USER error to refer to "installation
	  instructions" instead of long-gone httemplate/docs/install.html

2011-09-08 15:09  ivan

	* FS/FS/part_pkg/: delayed_Mixin.pm, discount_Mixin.pm, flat.pm,
	  global_Mixin.pm, prorate_Mixin.pm, sql_external.pm,
	  usage_Mixin.pm: fix mixin inheritence preventing prorate_delayed
	  packages from billing, RT#14372

2011-09-08 15:04  ivan

	* FS/FS/part_pkg.pm: fix very obsolote eval callbacks for
	  part_pkg.setup and recur, RT#14372

2011-08-31 01:26  mark

	* FS/FS/cust_pay.pm: fix payment application with term discounts,
	  #5318

2011-08-18 16:59  ivan

	* httemplate/search/: 477partVI_census.html, cust_pkg.cgi: fix
	  "multiple states" links, RT#13922

2011-08-18 16:56  ivan

	* httemplate/search/cust_pkg.cgi: attempt fixing "multiple states"
	  links, RT#13922

2011-08-18 16:50  ivan

	* httemplate/search/477partVI_census.html: attempt fixing "multiple
	  states" links, RT#13922

2011-08-18 16:43  ivan

	* FS/FS/cust_pkg.pm, httemplate/search/477partVI_census.html:
	  attempt fixing "multiple states" links, RT#13922

2011-08-18 16:39  ivan

	* FS/FS/cust_pkg.pm, httemplate/search/477partVI_census.html:
	  attempt fixing "multiple states" links, RT#13922

2011-08-18 16:29  ivan

	* FS/FS/cust_pkg.pm, httemplate/search/477partVI_census.html: fix
	  477 package links, RT#13922

2011-08-18 16:22  ivan

	* httemplate/search/477partVI_census.html: fix 477 package links
	  report_option arg, RT#13922

2011-08-18 16:14  ivan

	* httemplate/search/: cust_pkg.cgi, 477partVI_census.html: fix 477
	  package links report_option arg, RT#13922

2011-08-18 15:55  ivan

	* httemplate/search/elements/search-html.html: BR before "None
	  found" search message

2011-08-18 15:35  ivan

	* FS/FS/Report/FCC_477.pm: fix errors saving 477 form info if run
	  concurrently, RT#13922

2011-08-18 15:33  ivan

	* httemplate/search/477partVI_census.html: fix 477 package links
	  report_option arg, RT#13922

2011-08-18 15:11  ivan

	* httemplate/search/cust_pkg.cgi: fix 477 reporting w/multiple
	  classnums, RT#13922

2011-08-18 14:27  ivan

	* FS/FS/cust_pkg.pm, httemplate/search/477partVI_census.html,
	  httemplate/search/cust_pkg.cgi: fix 477 reporting w/multiple
	  classnums, RT#13922

2011-08-17 18:23  ivan

	* httemplate/search/477partVI_census.html: multiple state links in
	  FCC report, RT#13922

2011-08-17 18:15  ivan

	* FS/FS/Report/FCC_477.pm, FS/FS/cust_pkg.pm,
	  httemplate/search/477partVI_census.html,
	  httemplate/search/cust_pkg.cgi: multiple state links in FCC
	  report, RT#13922

2011-08-17 17:33  ivan

	* FS/FS/cust_pkg.pm, httemplate/search/cust_pkg.cgi: allow multiple
	  classnums in package search, RT#13922

2011-08-17 14:39  ivan

	* httemplate/search/477partVI_census.html: fix 477 censustract
	  links, add links for states, RT#13922

2011-08-17 14:07  ivan

	* httemplate/search/477partVI_census.html: fix 477 censustract
	  links, RT#13922

2011-08-15 21:49  ivan

	* FS/FS/pay_batch.pm: fix paymentech batch download, RT#7905

2011-08-15 19:09  mark

	* httemplate/: elements/progress-init.html,
	  misc/download-bill_batch.html, search/bill_batch.cgi: reliably
	  start batch download, #947

2011-08-15 18:10  ivan

	* httemplate/search/cust_main.cgi: fix package links in customer
	  search in 2.1, RT#14091

2011-08-15 17:44  ivan

	* conf/: invoice_html, invoice_latex: Invoice # & Customer #

2011-08-15 17:37  ivan

	* conf/: invoice_html, invoice_latex: don't show empty Terms: on
	  invoices

2011-08-15 13:51  ivan

	* httemplate/view/: cust_bill-pdf.cgi, cust_bill-ps.cgi: adding

FS/FS/Misc.pm
package FS::Misc;

use strict;
use vars qw ( @ISA @EXPORT_OK $DEBUG );
use Exporter;
use Carp;
use Data::Dumper;
use IPC::Run qw( run timeout );   # for _pslatex
use IPC::Run3; # for do_print... should just use IPC::Run i guess
use File::Temp;
use Tie::IxHash;
#do NOT depend on any FS:: modules here, causes weird (sometimes unreproducable
#until on client machine) dependancy loops.  put them in FS::Misc::Something
#instead

@ISA = qw( Exporter );
@EXPORT_OK = qw( generate_email send_email send_fax
                 states_hash counties state_label
                 card_types
                 pkg_freqs
                 generate_ps generate_pdf do_print
                 csv_from_fixed
               );

$DEBUG = 0;

=head1 NAME

FS::Misc - Miscellaneous subroutines

=head1 SYNOPSIS

  use FS::Misc qw(send_email);

  send_email();

=head1 DESCRIPTION

Miscellaneous subroutines.  This module contains miscellaneous subroutines
called from multiple other modules.  These are not OO or necessarily related,
but are collected here to elimiate code duplication.

=head1 SUBROUTINES

=over 4

=item generate_email OPTION => VALUE ...

Options:

=over 4

=item from

Sender address, required

=item to

Recipient address, required

=item bcc

Blind copy address, optional

=item subject

email subject, required

=item html_body

Email body (HTML alternative).  Arrayref of lines, or scalar.

Will be placed inside an HTML <BODY> tag.

=item text_body

Email body (Text alternative).  Arrayref of lines, or scalar.

=back

Returns an argument list to be passsed to L<send_email>.

=cut

#false laziness w/FS::cust_bill::generate_email

use MIME::Entity;
use HTML::Entities;

sub generate_email {
  my %args = @_;

  my $me = '[FS::Misc::generate_email]';

  my %return = (
    'from'    => $args{'from'},
    'to'      => $args{'to'},
    'bcc'     => $args{'bcc'},
    'subject' => $args{'subject'},
  );

  #if (ref($args{'to'}) eq 'ARRAY') {
  #  $return{'to'} = $args{'to'};
  #} else {
  #  $return{'to'} = [ grep { $_ !~ /^(POST|FAX)$/ }
  #                         $self->cust_main->invoicing_list
  #                  ];
  #}

  warn "$me creating HTML/text multipart message"
    if $DEBUG;

  $return{'nobody'} = 1;

  my $alternative = build MIME::Entity
    'Type'        => 'multipart/alternative',
    'Encoding'    => '7bit',
    'Disposition' => 'inline'
  ;

  my $data;
  if ( ref($args{'text_body'}) eq 'ARRAY' ) {
    $data = $args{'text_body'};
  } else {
    $data = [ split(/\n/, $args{'text_body'}) ];
  }

  $alternative->attach(
    'Type'        => 'text/plain',
    #'Encoding'    => 'quoted-printable',
    'Encoding'    => '7bit',
    'Data'        => $data,
    'Disposition' => 'inline',
  );

  my @html_data;
  if ( ref($args{'html_body'}) eq 'ARRAY' ) {
    @html_data = @{ $args{'html_body'} };
  } else {
    @html_data = split(/\n/, $args{'html_body'});
  }

  $alternative->attach(
    'Type'        => 'text/html',
    'Encoding'    => 'quoted-printable',
    'Data'        => [ '<html>',
                       '  <head>',
                       '    <title>',
                       '      '. encode_entities($return{'subject'}), 
                       '    </title>',
                       '  </head>',
                       '  <body bgcolor="#e8e8e8">',
                       @html_data,
                       '  </body>',
                       '</html>',
                     ],
    'Disposition' => 'inline',
    #'Filename'    => 'invoice.pdf',
  );

  #no other attachment:
  # multipart/related
  #   multipart/alternative
  #     text/plain
  #     text/html

  $return{'content-type'} = 'multipart/related';
  $return{'mimeparts'} = [ $alternative ];
  $return{'type'} = 'multipart/alternative'; #Content-Type of first part...
  #$return{'disposition'} = 'inline';

  %return;

}

=item send_email OPTION => VALUE ...

Options:

=over 4

=item from

(required)

=item to

(required) comma-separated scalar or arrayref of recipients

=item subject

(required)

=item content-type

(optional) MIME type for the body

=item body

(required unless I<nobody> is true) arrayref of body text lines

=item mimeparts

(optional, but required if I<nobody> is true) arrayref of MIME::Entity->build PARAMHASH refs or MIME::Entity objects.  These will be passed as arguments to MIME::Entity->attach().

=item nobody

(optional) when set true, send_email will ignore the I<body> option and simply construct a message with the given I<mimeparts>.  In this case,
I<content-type>, if specified, overrides the default "multipart/mixed" for the outermost MIME container.

=item content-encoding

(optional) when using nobody, optional top-level MIME
encoding which, if specified, overrides the default "7bit".

=item type

(optional) type parameter for multipart/related messages

=back

=cut

use vars qw( $conf );
use Date::Format;
use MIME::Entity;
use Email::Sender::Simple qw(sendmail);
use Email::Sender::Transport::SMTP;
use Email::Sender::Transport::SMTP::TLS;
use FS::UID;

FS::UID->install_callback( sub {
  $conf = new FS::Conf;
} );

sub send_email {
  my(%options) = @_;
  if ( $DEBUG ) {
    my %doptions = %options;
    $doptions{'body'} = '(full body not shown in debug)';
    warn "FS::Misc::send_email called with options:\n  ". Dumper(\%doptions);
#         join("\n", map { "  $_: ". $options{$_} } keys %options ). "\n"
  }

  my @to = ref($options{to}) ? @{ $options{to} } : ( $options{to} );

  my @mimeargs = ();
  my @mimeparts = ();
  if ( $options{'nobody'} ) {

    croak "'mimeparts' option required when 'nobody' option given\n"
      unless $options{'mimeparts'};

    @mimeparts = @{$options{'mimeparts'}};

    @mimeargs = (
      'Type'         => ( $options{'content-type'} || 'multipart/mixed' ),
      'Encoding'     => ( $options{'content-encoding'} || '7bit' ),
    );

  } else {

    @mimeparts = @{$options{'mimeparts'}}
      if ref($options{'mimeparts'}) eq 'ARRAY';

    if (scalar(@mimeparts)) {

      @mimeargs = (
        'Type'     => 'multipart/mixed',
        'Encoding' => '7bit',
      );
  
      unshift @mimeparts, { 
        'Type'        => ( $options{'content-type'} || 'text/plain' ),
        'Data'        => $options{'body'},
        'Encoding'    => ( $options{'content-type'} ? '-SUGGEST' : '7bit' ),
        'Disposition' => 'inline',
      };

    } else {
    
      @mimeargs = (
        'Type'     => ( $options{'content-type'} || 'text/plain' ),
        'Data'     => $options{'body'},
        'Encoding' => ( $options{'content-type'} ? '-SUGGEST' : '7bit' ),
      );

    }

  }

  my $domain;
  if ( $options{'from'} =~ /\@([\w\.\-]+)/ ) {
    $domain = $1;
  } else {
    warn 'no domain found in invoice from address '. $options{'from'}.
         '; constructing Message-ID (and saying HELO) @example.com'; 
    $domain = 'example.com';
  }
  my $message_id = join('.', rand()*(2**32), $$, time). "\@$domain";

  my $message = MIME::Entity->build(
    'From'       => $options{'from'},
    'To'         => join(', ', @to),
    'Sender'     => $options{'from'},
    'Reply-To'   => $options{'from'},
    'Date'       => time2str("%a, %d %b %Y %X %z", time),
    'Subject'    => $options{'subject'},
    'Message-ID' => "<$message_id>",
    @mimeargs,
  );

  if ( $options{'type'} ) {
    #false laziness w/cust_bill::generate_email
    $message->head->replace('Content-type',
      $message->mime_type.
      '; boundary="'. $message->head->multipart_boundary. '"'.
      '; type='. $options{'type'}
    );
  }

  foreach my $part (@mimeparts) {

    if ( UNIVERSAL::isa($part, 'MIME::Entity') ) {

      warn "attaching MIME part from MIME::Entity object\n"
        if $DEBUG;
      $message->add_part($part);

    } elsif ( ref($part) eq 'HASH' ) {

      warn "attaching MIME part from hashref:\n".
           join("\n", map "  $_: ".$part->{$_}, keys %$part ). "\n"
        if $DEBUG;
      $message->attach(%$part);

    } else {
      croak "mimepart $part isn't a hashref or MIME::Entity object!";
    }

  }

  #send the email

  my %smtp_opt = ( 'host' => $conf->config('smtpmachine'),
                   'helo' => $domain,
                 );

  my($port, $enc) = split('-', ($conf->config('smtp-encryption') || '25') );
  $smtp_opt{'port'} = $port;

  my $transport;
  if ( defined($enc) && $enc eq 'starttls' ) {
    $smtp_opt{$_} = $conf->config("smtp-$_") for qw(username password);
    $transport = Email::Sender::Transport::SMTP::TLS->new( %smtp_opt );
  } else {
    if ( $conf->exists('smtp-username') && $conf->exists('smtp-password') ) {
      $smtp_opt{"sasl_$_"} = $conf->config("smtp-$_") for qw(username password);
    }
    $smtp_opt{'ssl'} = 1 if defined($enc) && $enc eq 'tls';
    $transport = Email::Sender::Transport::SMTP->new( %smtp_opt );
  }
 
  push @to, $options{bcc} if defined($options{bcc});
  local $@; # just in case
  eval { sendmail($message, { transport => $transport,
                              from      => $options{from},
                              to        => \@to }) };
 
  if(ref($@) and $@->isa('Email::Sender::Failure')) {
    return ($@->code ? $@->code.' ' : '').$@->message
  }
  else {
    return $@;
  }
}

=item send_fax OPTION => VALUE ...

Options:

I<dialstring> - (required) 10-digit phone number w/ area code

I<docdata> - (required) Array ref containing PostScript or TIFF Class F document

-or-

I<docfile> - (required) Filename of PostScript TIFF Class F document

...any other options will be passed to L<Fax::Hylafax::Client::sendfax>


=cut

sub send_fax {

  my %options = @_;

  die 'HylaFAX support has not been configured.'
    unless $conf->exists('hylafax');

  eval {
    require Fax::Hylafax::Client;
  };

  if ($@) {
    if ($@ =~ /^Can't locate Fax.*/) {
      die "You must have Fax::Hylafax::Client installed to use invoice faxing."
    } else {
      die $@;
    }
  }

  my %hylafax_opts = map { split /\s+/ } $conf->config('hylafax');

  die 'Called send_fax without a \'dialstring\'.'
    unless exists($options{'dialstring'});

  if (exists($options{'docdata'}) and ref($options{'docdata'}) eq 'ARRAY') {
      my $dir = $FS::UID::conf_dir. "/cache.". $FS::UID::datasrc;
      my $fh = new File::Temp(
        TEMPLATE => 'faxdoc.'. $options{'dialstring'} . '.XXXXXXXX',
        DIR      => $dir,
        UNLINK   => 0,
      ) or die "can't open temp file: $!\n";

      $options{docfile} = $fh->filename;

      print $fh @{$options{'docdata'}};
      close $fh;

      delete $options{'docdata'};
  }

  die 'Called send_fax without a \'docfile\' or \'docdata\'.'
    unless exists($options{'docfile'});

  #FIXME: Need to send canonical dialstring to HylaFAX, but this only
  #       works in the US.

  $options{'dialstring'} =~ s/[^\d\+]//g;
  if ($options{'dialstring'} =~ /^\d{10}$/) {
    $options{dialstring} = '+1' . $options{'dialstring'};
  } else {
    return 'Invalid dialstring ' . $options{'dialstring'} . '.';
  }

  my $faxjob = &Fax::Hylafax::Client::sendfax(%options, %hylafax_opts);

  if ($faxjob->success) {
    warn "Successfully queued fax to '$options{dialstring}' with jobid " .
           $faxjob->jobid
      if $DEBUG;
    return '';
  } else {
    return 'Error while sending FAX: ' . $faxjob->trace;
  }

}

=item states_hash COUNTRY

Returns a list of key/value pairs containing state (or other sub-country
division) abbriviations and names.

=cut

use FS::Record qw(qsearch);
use Locale::SubCountry;

sub states_hash {
  my($country) = @_;

  my @states = 
#     sort
     map { s/[\n\r]//g; $_; }
     map { $_->state; }
         qsearch({ 
                   'select'    => 'state',
                   'table'     => 'cust_main_county',
                   'hashref'   => { 'country' => $country },
                   'extra_sql' => 'GROUP BY state',
                });

  #it could throw a fatal "Invalid country code" error (for example "AX")
  my $subcountry = eval { new Locale::SubCountry($country) }
    or return ( '', '(n/a)' );

  #"i see your schwartz is as big as mine!"
  map  { ( $_->[0] => $_->[1] ) }
  sort { $a->[1] cmp $b->[1] }
  map  { [ $_ => state_label($_, $subcountry) ] }
       @states;
}

=item counties STATE COUNTRY

Returns a list of counties for this state and country.

=cut

sub counties {
  my( $state, $country ) = @_;

  sort map { s/[\n\r]//g; $_; }
       map { $_->county }
           qsearch({
             'select'  => 'DISTINCT county',
             'table'   => 'cust_main_county',
             'hashref' => { 'state'   => $state,
                            'country' => $country,
                          },
           });
}

=item state_label STATE COUNTRY_OR_LOCALE_SUBCOUNRY_OBJECT

=cut

sub state_label {
  my( $state, $country ) = @_;

  unless ( ref($country) ) {
    $country = eval { new Locale::SubCountry($country) }
      or return'(n/a)';

  }

  # US kludge to avoid changing existing behaviour 
  # also we actually *use* the abbriviations...
  my $full_name = $country->country_code eq 'US'
                    ? ''
                    : $country->full_name($state);

  $full_name = '' if $full_name eq 'unknown';
  $full_name =~ s/\(see also.*\)\s*$//;
  $full_name .= " ($state)" if $full_name;

  $full_name || $state || '(n/a)';

}

=item card_types

Returns a hash reference of the accepted credit card types.  Keys are shorter
identifiers and values are the longer strings used by the system (see
L<Business::CreditCard>).

=cut

#$conf from above

sub card_types {
  my $conf = new FS::Conf;

  my %card_types = (
    #displayname                    #value (Business::CreditCard)
    "VISA"                       => "VISA card",
    "MasterCard"                 => "MasterCard",
    "Discover"                   => "Discover card",
    "American Express"           => "American Express card",
    "Diner's Club/Carte Blanche" => "Diner's Club/Carte Blanche",
    "enRoute"                    => "enRoute",
    "JCB"                        => "JCB",
    "BankCard"                   => "BankCard",
    "Switch"                     => "Switch",
    "Solo"                       => "Solo",
  );
  my @conf_card_types = grep { ! /^\s*$/ } $conf->config('card-types');
  if ( @conf_card_types ) {
    #perhaps the hash is backwards for this, but this way works better for
    #usage in selfservice
    %card_types = map  { $_ => $card_types{$_} }
                  grep {
                         my $d = $_;
			   grep { $card_types{$d} eq $_ } @conf_card_types
                       }
		    keys %card_types;
  }

  \%card_types;
}

=item generate_ps FILENAME

Returns an postscript rendition of the LaTex file, as a scalar.
FILENAME does not contain the .tex suffix and is unlinked by this function.

=cut

use String::ShellQuote;

sub generate_ps {
  my $file = shift;

  my $dir = $FS::UID::conf_dir. "/cache.". $FS::UID::datasrc;
  chdir($dir);

  _pslatex($file);

  system('dvips', '-q', '-t', 'letter', "$file.dvi", '-o', "$file.ps" ) == 0
    or die "dvips failed";

  open(POSTSCRIPT, "<$file.ps")
    or die "can't open $file.ps: $! (error in LaTeX template?)\n";

  unlink("$file.dvi", "$file.log", "$file.aux", "$file.ps", "$file.tex");

  my $ps = '';

  if ( $conf->exists('lpr-postscript_prefix') ) {
    my $prefix = $conf->config('lpr-postscript_prefix');
    $ps .= eval qq("$prefix");
  }

  while (<POSTSCRIPT>) {
    $ps .= $_;
  }

  close POSTSCRIPT;

  if ( $conf->exists('lpr-postscript_suffix') ) {
    my $suffix = $conf->config('lpr-postscript_suffix');
    $ps .= eval qq("$suffix");
  }

  return $ps;

}

=item pkg_freqs

Returns a hash reference of allowed package billing frequencies.

=cut

sub pkg_freqs {
  tie my %freq, 'Tie::IxHash', (
    '0'    => '(no recurring fee)',
    '1h'   => 'hourly',
    '1d'   => 'daily',
    '2d'   => 'every two days',
    '3d'   => 'every three days',
    '1w'   => 'weekly',
    '2w'   => 'biweekly (every 2 weeks)',
    '1'    => 'monthly',
    '45d'  => 'every 45 days',
    '2'    => 'bimonthly (every 2 months)',
    '3'    => 'quarterly (every 3 months)',
    '4'    => 'every 4 months',
    '137d' => 'every 4 1/2 months (137 days)',
    '6'    => 'semiannually (every 6 months)',
    '12'   => 'annually',
    '13'   => 'every 13 months (annually +1 month)',
    '24'   => 'biannually (every 2 years)',
    '36'   => 'triannually (every 3 years)',
    '48'   => '(every 4 years)',
    '60'   => '(every 5 years)',
    '120'  => '(every 10 years)',
  ) ;
  \%freq;
}

=item generate_pdf FILENAME

Returns an PDF rendition of the LaTex file, as a scalar.  FILENAME does not
contain the .tex suffix and is unlinked by this function.

=cut

use String::ShellQuote;

sub generate_pdf {
  my $file = shift;

  my $dir = $FS::UID::conf_dir. "/cache.". $FS::UID::datasrc;
  chdir($dir);

  #system('pdflatex', "$file.tex");
  #system('pdflatex', "$file.tex");
  #! LaTeX Error: Unknown graphics extension: .eps.

  _pslatex($file);

  my $sfile = shell_quote $file;

  #system('dvipdf', "$file.dvi", "$file.pdf" );
  system(
    "dvips -q -t letter -f $sfile.dvi ".
    "| gs -q -dNOPAUSE -dBATCH -sDEVICE=pdfwrite -sOutputFile=$sfile.pdf ".
    "     -c save pop -"
  ) == 0
    or die "dvips | gs failed: $!";

  open(PDF, "<$file.pdf")
    or die "can't open $file.pdf: $! (error in LaTeX template?)\n";

  unlink("$file.dvi", "$file.log", "$file.aux", "$file.pdf", "$file.tex");

  my $pdf = '';
  while (<PDF>) {
    $pdf .= $_;
  }

  close PDF;

  return $pdf;

}

sub _pslatex {
  my $file = shift;

  #my $sfile = shell_quote $file;

  my @cmd = (
    'latex',
    '-interaction=batchmode',
    '\AtBeginDocument{\RequirePackage{pslatex}}',
    '\def\PSLATEXTMP{\futurelet\PSLATEXTMP\PSLATEXTMPB}',
    '\def\PSLATEXTMPB{\ifx\PSLATEXTMP\nonstopmode\else\input\fi}',
    '\PSLATEXTMP',
    "$file.tex"
  );

  my $timeout = 30; #? should be more than enough

  for ( 1, 2 ) {

    local($SIG{CHLD}) = sub {};
    run( \@cmd, '>'=>'/dev/null', '2>'=>'/dev/null', timeout($timeout) )
      or die "pslatex $file.tex failed; see $file.log for details?\n";

  }

}

=item print ARRAYREF

Sends the lines in ARRAYREF to the printer.

=cut

sub do_print {
  my $data = shift;

  my $lpr = $conf->config('lpr');

  my $outerr = '';
  run3 $lpr, $data, \$outerr, \$outerr;
  if ( $? ) {
    $outerr = ": $outerr" if length($outerr);
    die "Error from $lpr (exit status ". ($?>>8). ")$outerr\n";
  }

}

=item csv_from_fixed, FILEREF COUNTREF, [ LENGTH_LISTREF, [ CALLBACKS_LISTREF ] ]

Converts the filehandle referenced by FILEREF from fixed length record
lines to a CSV file according to the lengths specified in LENGTH_LISTREF.
The CALLBACKS_LISTREF refers to a correpsonding list of coderefs.  Each
should return the value to be substituted in place of its single argument.

Returns false on success or an error if one occurs.

=cut

sub csv_from_fixed {
  my( $fhref, $countref, $lengths, $callbacks) = @_;

  eval { require Text::CSV_XS; };
  return $@ if $@;

  my $ofh = $$fhref;
  my $unpacker = new Text::CSV_XS;
  my $total = 0;
  my $template = join('', map {$total += $_; "A$_"} @$lengths) if $lengths;

  my $dir = "%%%FREESIDE_CACHE%%%/cache.$FS::UID::datasrc";
  my $fh = new File::Temp( TEMPLATE => "FILE.csv.XXXXXXXX",
                           DIR      => $dir,
                           UNLINK   => 0,
                         ) or return "can't open temp file: $!\n"
    if $template;

  while ( defined(my $line=<$ofh>) ) {
    $$countref++;
    if ( $template ) {
      my $column = 0;

      chomp $line;
      return "unexpected input at line $$countref: $line".
             " -- expected $total but received ". length($line)
        unless length($line) == $total;

      $unpacker->combine( map { my $i = $column++;
                                defined( $callbacks->[$i] )
                                  ? &{ $callbacks->[$i] }( $_ )
                                  : $_
                              } unpack( $template, $line )
                        )
        or return "invalid data for CSV: ". $unpacker->error_input;

      print $fh $unpacker->string(), "\n"
        or return "can't write temp file: $!\n";
    }
  }

  if ( $template ) { close $$fhref; $$fhref = $fh }

  seek $$fhref, 0, 0;
  '';
}


=back

=head1 BUGS

This package exists.

=head1 SEE ALSO

L<FS::UID>, L<FS::CGI>, L<FS::Record>, the base documentation.

L<Fax::Hylafax::Client>

=cut

1;
